packages/connector/src/services/core/controllers/stream.ts

```diff
@@ -1,39 +1,46 @@
-import * as assert from 'assert'
-import { StreamServer } from '@interledger/stream-receiver'
 import { isIlpReply } from 'ilp-packet'
 import { RafikiContext, RafikiMiddleware } from '../rafiki'
 
-interface StreamControllerOptions {
-  serverSecret: Buffer
-  serverAddress: string
-}
+const CONNECTION_EXPIRY = 60 * 60 // seconds
 
-export function createStreamController({
-  serverSecret,
-  serverAddress
-}: StreamControllerOptions): RafikiMiddleware {
-  assert.equal(serverSecret.length, 32)
-  const server = new StreamServer({ serverSecret, serverAddress })
+// Track the total amount received per stream connection.
+export const streamReceivedKey = (connectionId: string): string =>
+  `stream_received:${connectionId}`
 
+export function createStreamController(): RafikiMiddleware {
   return async function (
     ctx: RafikiContext,
     next: () => Promise<unknown>
   ): Promise<void> {
+    const { redis, streamServer } = ctx.services
     const { request, response } = ctx
 
     const { stream } = ctx.accounts.outgoing
     if (
       !stream ||
       !stream.enabled ||
-      !server.decodePaymentTag(request.prepare.destination)
+      !streamServer.decodePaymentTag(request.prepare.destination) // XXX mark this earlier in the middleware pipeline
     ) {
       await next()
       return
     }
 
-    const moneyOrReply = server.createReply(request.prepare)
-    response.reply = isIlpReply(moneyOrReply)
-      ? moneyOrReply
-      : moneyOrReply.accept()
+    const moneyOrReply = streamServer.createReply(request.prepare)
+    if (isIlpReply(moneyOrReply)) {
+      response.reply = moneyOrReply
+      return
+    }
+
+    const { connectionId } = moneyOrReply
+    const totalReceived = await redis.incrby(
+      streamReceivedKey(connectionId),
+      +request.prepare.amount
+    )
+    moneyOrReply.setTotalReceived(totalReceived)
+    if (totalReceived !== +request.prepare.amount) {
+      // Set key expiry once, when the first packet arrives.
+      await redis.expire(connectionId, CONNECTION_EXPIRY)
```

Comment:
Would it be more desirable set the key expiry on every packet, to reset/extend it as new packets come in?

Reply:
Yeah, that would work. I figured this way would have fewer redis requests, but `expire`-ing after every increment would probably be more accurate.

Reply:
Or pipeline it with the `incrby`?

Reply:
I'm not sure what you mean by pipeline -- `incrby` doesn't have an extra `expiry parameter like some commands.

Reply:
I think it'd look something like this:
```js
redis.pipeline()
  .incrby(...)
  .expire(...)
  .exec()
```
So ioredis sends both commands in a single batch instead of multiple network calls: https://github.com/luin/ioredis#pipelining

Reply:
Interesting, I didn't know about that. :+1:

Reply:
I ended up just using separate commands -- that link suggests pipelining for batches of >5 commands. Also, there's an option called `enableAutoPipelining` that we could enable eventually,  so I don't think its worth explicitly pipelining commands in the code.

Reply:
`enableAutoPipelining` -- possibly, but it looks like this might require calling all the functions in the same tick which might not work well with async/await.

It still seems like this is adding unnecessary latency. Could you at least `Promise.all` the `incrby` and `expire` calls?

And could you use the `stringNumbers` config option so `incrby` returns a string instead of requiring a sequential `get` call?

Reply:
https://github.com/luin/ioredis/blob/abd9a82433ad67b91a4bddb45aff8da2e20d45e8/API.md says it batched by `event loop iteration` -- I don't think that's the same thing as a tick. Either way, I suspect it's a premature optimization.

`Promise.all` could lead to the `expire` being processed before `incrby`. If that happens on the first packet, no expiry will be set. What if it just didn't `await` the `expire` command?

:+1: for `stringNumbers`, I'll add that.

Reply:
Oh, good point. Then this seems like a good use case for a transaction using `multi`? Then Redis ensures they're executed sequentially, and we don't need to wait for the response before issuing the next command